### ISSUE TITLE
Empty search term gives exception

### ISSUE THREAD

**Issue by luukschakenraad** [EmicoEcommerce/Magento2Tweakwise-archived#67]
On mobile, it's easy for a user to press "search" without entering a search parameter. The user is redirected to `/catalogsearch/result/?q=`. This page gives an exception.

Environment
--
* PHP Version: 7.1
* Magento Version: 2.2.7
* Tweakwise Version: 1.4.9
* Tweakwise Export Version: 1.3.3
* Magento Deploy Mode: production, development

Steps to reproduce
--
1. Install Magento with Tweakwise module and set correct config
2. Visit `/catalogsearch/result/?q=`

Actual result
--
Exception:
```
1 exception(s):
Exception #0 (Emico\Tweakwise\Exception\ApiException): Invalid response received by Tweakwise server, response code is not 200. Request "http://navigator-group1.tweakwise.com/navigation-search/e6c5a413?tn_ps=24&tn_fk_visibility=3%7C4&tn_ft=8"

Exception #0 (Emico\Tweakwise\Exception\ApiException): Invalid response received by Tweakwise server, response code is not 200. Request "http://navigator-group1.tweakwise.com/navigation-search/e6c5a413?tn_ps=24&tn_fk_visibility=3%7C4&tn_ft=8"
#0 /magento/vendor/emico/tweakwise/src/Model/Client.php(166): Emico\Tweakwise\Model\Client->doRequest(Object(Emico\Tweakwise\Model\Client\Request\ProductSearchRequest))
#1 /magento/vendor/emico/tweakwise/src/Model/Catalog/Layer/NavigationContext.php(145): Emico\Tweakwise\Model\Client->request(Object(Emico\Tweakwise\Model\Client\Request\ProductSearchRequest))
#2 /magento/vendor/emico/tweakwise/src/Model/Catalog/Layer/NavigationContext/CurrentContext.php(59): Emico\Tweakwise\Model\Catalog\Layer\NavigationContext->getResponse()
#3 /magento/vendor/emico/tweakwise/src/Model/Observer/CatalogSearchRedirect.php(32): Emico\Tweakwise\Model\Catalog\Layer\NavigationContext\CurrentContext->getResponse()
#4 /magento/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(111): Emico\Tweakwise\Model\Observer\CatalogSearchRedirect->__construct(Object(Emico\Tweakwise\Model\Config), Object(Emico\Tweakwise\Model\Catalog\Layer\NavigationContext\CurrentContext), Object(Magento\Framework\App\Action\Context), Object(Emico\Tweakwise\Model\Catalog\Layer\NavigationContext))
#5 /magento/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(66): Magento\Framework\ObjectManager\Factory\AbstractFactory->createObject('Emico\\Tweakwise...', Array)
#6 /magento/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Emico\\Tweakwise...')
#7 /magento/vendor/magento/framework/Event/ObserverFactory.php(33): Magento\Framework\ObjectManager\ObjectManager->get('Emico\\Tweakwise...')
#8 /magento/vendor/magento/framework/Event/Invoker/InvokerDefault.php(58): Magento\Framework\Event\ObserverFactory->get('Emico\\Tweakwise...')
#9 /magento/vendor/magento/module-staging/Model/Event/Manager.php(97): Magento\Framework\Event\Invoker\InvokerDefault->dispatch(Array, Object(Magento\Framework\Event\Observer))
#10 /magento/generated/code/Magento/Staging/Model/Event/Manager/Proxy.php(95): Magento\Staging\Model\Event\Manager->dispatch('controller_acti...', Array)
#11 /magento/vendor/magento/framework/App/Action/Action.php(100): Magento\Staging\Model\Event\Manager\Proxy->dispatch('controller_acti...', Array)
#12 /magento/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\Action\Action->dispatch(Object(Magento\Framework\App\Request\Http))
#13 /magento/vendor/magento/framework/Interception/Interceptor.php(138): Magento\CatalogSearch\Controller\Result\Index\Interceptor->___callParent('dispatch', Array)
#14 /magento/vendor/magento/module-customer-segment/Model/App/Action/ContextPlugin.php(81): Magento\CatalogSearch\Controller\Result\Index\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#15 /magento/vendor/magento/framework/Interception/Interceptor.php(135): Magento\CustomerSegment\Model\App\Action\ContextPlugin->aroundDispatch(Object(Magento\CatalogSearch\Controller\Result\Index\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#16 /magento/vendor/magento/framework/Interception/Interceptor.php(153): Magento\CatalogSearch\Controller\Result\Index\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#17 /magento/generated/code/Magento/CatalogSearch/Controller/Result/Index/Interceptor.php(39): Magento\CatalogSearch\Controller\Result\Index\Interceptor->___callPlugins('dispatch', Array, Array)
#18 /magento/vendor/magento/framework/App/FrontController.php(55): Magento\CatalogSearch\Controller\Result\Index\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#19 /magento/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\FrontController->dispatch(Object(Magento\Framework\App\Request\Http))
#20 /magento/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\FrontController\Interceptor->___callParent('dispatch', Array)
#21 /magento/vendor/magento/module-store/App/FrontController/Plugin/RequestPreprocessor.php(94): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#22 /magento/vendor/magento/framework/Interception/Interceptor.php(135): Magento\Store\App\FrontController\Plugin\RequestPreprocessor->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#23 /magento/vendor/magento/module-page-cache/Model/App/FrontController/BuiltinPlugin.php(69): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#24 /magento/vendor/magento/framework/Interception/Interceptor.php(135): Magento\PageCache\Model\App\FrontController\BuiltinPlugin->aroundDispatch(Object(Magento\Framework\App\FrontController\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#25 /magento/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#26 /magento/generated/code/Magento/Framework/App/FrontController/Interceptor.php(26): Magento\Framework\App\FrontController\Interceptor->___callPlugins('dispatch', Array, NULL)
#27 /magento/vendor/magento/framework/App/Http.php(135): Magento\Framework\App\FrontController\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#28 /magento/generated/code/Magento/Framework/App/Http/Interceptor.php(24): Magento\Framework\App\Http->launch()
#29 /magento/vendor/magento/framework/App/Bootstrap.php(257): Magento\Framework\App\Http\Interceptor->launch()
#30 /magento/pub/index.php(37): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
#31 /Users/luuk/.composer/vendor/weprovide/valet-plus/server.php(131): require('/Volumes/CaseSe...')
#32 {main}
```

```
<error xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
<message>no (correct) searchterm provided</message>
</error>
```

Expected result
--
Redirect to homepage (default Magento)

Possible fix
--
Replace in file `magento/vendor/emico/tweakwise/src/Model/Catalog/Layer/Url/AbstractUrl.php`:

```
/**
     * @param HttpRequest $request
     * @return string|null
     */
    protected function getSearch(HttpRequest $request)
    {
        return $request->getQuery(self::PARAM_SEARCH);
    }
```

With:

```
/**
     * @param HttpRequest $request
     * @return string|null
     */
    protected function getSearch(HttpRequest $request)
    {
        return $request->getQuery(self::PARAM_SEARCH) ?: 'false';
    }
```

**Comment by edwinljacobs** [comment]
Hi,

Thanks for reporting this issue, I just created a branch to fix this issue (issue/67-exception-empty-search). We choose to let Magento handle the empty parameter instead of applying some default param which would lead to unexpected results. Could you verify that using said branch fixes the issue for you? If so I will merge it soon.

With kind regards!

**Comment by luukschakenraad** [comment]
Thx, @edwinljacobs. Good solution. I've just verified it and it works as expected.

**Comment by edwinljacobs** [comment]
Ah nice! I will release it somewhere next week.

**Comment by edwinljacobs** [comment]
Branch has been merged and is ready for release

**Comment by edwinljacobs** [comment]
Has been released in version v1.4.10

**Comment by luukschakenraad** [comment]
Thank you, @edwinljacobs
